### PR TITLE
Recover user settings

### PR DESCRIPTION
**Pull request by joseivanlopez**
## Problem

Last used username and room are lost from the login form after refreshing.

* https://github.com/jangouts/jangouts/issues/359

## Solution

Introduce `UserSettings` class, which stores and retrieves its data from browser local storage.

### CODE REVIEW

File: src/components/LoginForm/LoginForm.js [hunk 1]
```diff
@@ -45,8 +45,11 @@ function findRoom(rooms, roomId) {
  * @param {Function} dispatch - the dispatch function available in the Redux store
  * @returns {Function}
  */
-function onSubmit(dispatch) {
+function onSubmit(dispatch, settings) {
   return function(data) {
+    settings.username = data.username;
+    settings.roomId = data.room;
+    dispatch(roomActions.saveSettings(settings));
     dispatch(roomActions.login(data.username, data.room, data.pin));
```
Comment (dgdavid) [inline src/components/LoginForm/LoginForm.js:53]: JFYI: I have learned that React Redux added [`batch()`](https://react-redux.js.org/api/batch) to its API in version 7, which 

>  allows any React updates in an event loop tick to be batched together into a single render pas

However, it is useful only when using React < 18. Hopefully, we could upgrade the project to the latest React version soon.

Comment (joseivanlopez) [reply src/components/LoginForm/LoginForm.js:55]: Thanks for the info ;)

File: src/components/LoginForm/LoginForm.js [hunk 2]
```diff
@@ -47,8 +47,11 @@ function findRoom(rooms, roomId) {
  * @param {Function} dispatch - the dispatch function available in the Redux store
  * @returns {Function}
  */
-function onSubmit(dispatch) {
+function onSubmit(dispatch, settings) {
   return function(data) {
+    settings.username = data.username;
```
Comment (imobachgs) [inline src/components/LoginForm/LoginForm.js:52]: I would avoid mutating the `settings` object. After all, there is no reason to do that (and you are modifying an input argument). What about something like? (disclaimer: haven't tried :wink:).

```javascript
dispatch(roomActions.saveSettings({...settings, username: data.username, roomdId: data.room}).
```

File: src/components/LoginForm/LoginForm.js [hunk 3]
```diff
@@ -57,55 +60,49 @@ function onSubmit(dispatch) {
...
-      if (previousRoom.roomId) {
-        roomId = previousRoom.roomId;
-      } else if (rooms.length > 0) {
-        roomId = rooms[0].id;
-      }
-      reset({ room: roomId });
-      setSelectedRoom(findRoom(rooms, roomId));
-    });
+    dispatch(roomActions.loadSettings());
+    janusApi.getRooms().then(setRooms);
```
Comment (imobachgs) [inline src/components/LoginForm/LoginForm.js:94]: Not for this PR, but in the future we should store the rooms in the Redux store too.

File: src/components/LoginForm/LoginForm.js [hunk 4]
```diff
@@ -57,55 +60,49 @@ function onSubmit(dispatch) {
...
-          <svg
-            className="fill-current h-6 w-6 text-secondary mr-4"
-            xmlns="http://www.w3.org/2000/svg"
-            viewBox="0 0 20 20">
-            <path d="M2.93 17.07A10 10 0 1 1 17.07 2.93 10 10 0 0 1 2.93 17.07zm12.73-1.41A8 8 0 1 0 4.34 4.34a8 8 0 0 0 11.32 11.32zM9 11V9h2v6H9v-4zm0-6h2v2H9V5z" />
-          </svg>
-          <div>
-            <p className="font-bold">{error}</p>
-          </div>
+  return (
```
Comment (imobachgs) [inline src/components/LoginForm/LoginForm.js:66]: :+1:

File: src/components/LoginForm/LoginForm.js [hunk 5]
```diff
@@ -57,55 +60,49 @@ function onSubmit(dispatch) {
...
 }
 
 function LoginForm() {
   const dispatch = useDispatch();
   const [rooms, setRooms] = useState([]);
   const [selectedRoom, setSelectedRoom] = useState({});
-  const previousRoom = useSelector((state) => state.room);
   const { register, handleSubmit, reset } = useForm();
-  const { username: previousUsername, error } = previousRoom;
+  const { settings, error } = useSelector((state) => state.room);
```
Comment (imobachgs) [inline src/components/LoginForm/LoginForm.js:90]: If you want, you can destructure the `settings` object into `username` and `roomId`. As you prefer.

File: src/state/ducks/room.js [hunk 6]
```diff
@@ -50,26 +52,41 @@ const logout = () => {
   };
 };
 
+const loadSettings = () => {
+  const settings = UserSettings.load();
+
```
Comment (imobachgs) [inline src/state/ducks/room.js:57]: Why those empty lines?

File: src/state/ducks/room.test.js [hunk 7]
```diff
@@ -48,6 +49,18 @@ it('handles ROOM_LOGIN_REQUEST', () => {
   expect(reducer({ roomId, username }, action)).toEqual({ roomId, username, loggingIn: true });
 });
 
+it('handles ROOM_SETTINGS_LOAD', () => {
+  const settings = new UserSettings();
+  settings.username = "test";
+
+  const action = {
+    type: types.ROOM_SETTINGS_LOAD,
+    payload: { settings: settings }
```
Comment (imobachgs) [inline src/state/ducks/room.test.js:58]: You can do just `{ settings }`.

**Review by imobachgs — CHANGES_REQUESTED** [review]
In general, it looks good, thanks :-). The only important part, IMHO, it is the handling of the `settings` argument in the `onSubmit` function.

**Review by imobachgs — APPROVED** [review]

**Review by imobachgs — APPROVED** [review]

**Pull request merged**